### ISSUE TITLE
ClassCastException in AbstractGemFireOperationsSessionRepository.afterCreate

### ISSUE THREAD

**Issue by sshcherbakov** [spring-projects/spring-session#373]
I'm trying to setup a simple Spring security session sharing using spring-session and Gemfire.
I've added following dependencies to my webapp project:
````
		compile('org.springframework.session:spring-session-data-gemfire:1.1.0.RC1')
		compile('org.springframework.session:spring-session:1.1.0.RC1')
		compile('org.springframework.security:spring-security-web')
		compile('org.springframework.security:spring-security-config')
````
and almost succeeded to setup session sharing. 
I had to add several Spring dependencies to the Gemfire classpath because the spring-session-data-gemfire doesn't use PDX and thus some classes need to be present on the server side as well:
`spring-session-1.1.0.RC1.jar`,`spring-expression-4.2.4.RELEASE.jar`, `spring-security-web-4.0.3.RELEASE.jar`, `spring-security-core-4.0.3.RELEASE.jar`
The session gets stored in my Gemfire cluster and I don't get server side ClassNotFoundExceptions any more.
I also get a simple authentication prompt asking to login to my application:
````
@EnableWebSecurity
public class SecurityConfig extends WebSecurityConfigurerAdapter {

	@Autowired
	public void configureGlobal(AuthenticationManagerBuilder auth) throws Exception {
		auth
			.inMemoryAuthentication()
				.withUser("user").password("password").roles("USER");
	}
}
````
On the application side after I try to login with the those credentials I get the following exception:
````
[error 2016/02/22 02:29:59.823 CET <http-nio-8080-exec-2> tid=0x31] Exception occurred in CacheListener
java.lang.ClassCastException: com.gemstone.gemfire.internal.cache.Token$Tombstone cannot be cast to org.springframework.session.ExpiringSession
	at org.springframework.session.data.gemfire.AbstractGemFireOperationsSessionRepository.afterCreate(AbstractGemFireOperationsSessionRepository.java:212)
	at com.gemstone.gemfire.internal.cache.EnumListenerEvent$AFTER_CREATE.dispatchEvent(EnumListenerEvent.java:97)
	at com.gemstone.gemfire.internal.cache.LocalRegion.dispatchEvent(LocalRegion.java:8820)
	at com.gemstone.gemfire.internal.cache.LocalRegion.dispatchListenerEvent(LocalRegion.java:7311)
	at com.gemstone.gemfire.internal.cache.LocalRegion.invokePutCallbacks(LocalRegion.java:6133)
	at com.gemstone.gemfire.internal.cache.EntryEventImpl.invokeCallbacks(EntryEventImpl.java:1919)
	at com.gemstone.gemfire.internal.cache.ProxyRegionMap$ProxyRegionEntry.dispatchListenerEvents(ProxyRegionMap.java:544)
	at com.gemstone.gemfire.internal.cache.LocalRegion.basicPutPart2(LocalRegion.java:5987)
	at com.gemstone.gemfire.internal.cache.ProxyRegionMap.basicPut(ProxyRegionMap.java:228)
	at com.gemstone.gemfire.internal.cache.LocalRegion.basicPutEntry(LocalRegion.java:5919)
	at com.gemstone.gemfire.internal.cache.LocalRegion.findObjectInSystem(LocalRegion.java:2808)
	at com.gemstone.gemfire.internal.cache.LocalRegion.nonTxnFindObject(LocalRegion.java:1464)
	at com.gemstone.gemfire.internal.cache.LocalRegionDataView.findObject(LocalRegionDataView.java:133)
	at com.gemstone.gemfire.internal.cache.LocalRegion.get(LocalRegion.java:1348)
	at com.gemstone.gemfire.internal.cache.LocalRegion.get(LocalRegion.java:1311)
	at com.gemstone.gemfire.internal.cache.LocalRegion.get(LocalRegion.java:1298)
	at com.gemstone.gemfire.internal.cache.AbstractRegion.get(AbstractRegion.java:321)
	at org.springframework.data.gemfire.GemfireTemplate.get(GemfireTemplate.java:161)
	at org.springframework.session.data.gemfire.GemFireOperationsSessionRepository.getSession(GemFireOperationsSessionRepository.java:117)
	at org.springframework.session.data.gemfire.GemFireOperationsSessionRepository.getSession(GemFireOperationsSessionRepository.java:38)
	at org.springframework.session.web.http.SessionRepositoryFilter$SessionRepositoryRequestWrapper.getSession(SessionRepositoryFilter.java:270)
	at org.springframework.session.web.http.SessionRepositoryFilter$SessionRepositoryRequestWrapper.getSession(SessionRepositoryFilter.java:286)
	at org.springframework.session.web.http.SessionRepositoryFilter$SessionRepositoryRequestWrapper.getSession(SessionRepositoryFilter.java:170)
	at javax.servlet.http.HttpServletRequestWrapper.getSession(HttpServletRequestWrapper.java:231)
	at javax.servlet.http.HttpServletRequestWrapper.getSession(HttpServletRequestWrapper.java:231)
	at org.springframework.security.web.csrf.HttpSessionCsrfTokenRepository.loadToken(HttpSessionCsrfTokenRepository.java:77)
	at org.springframework.security.web.csrf.CsrfFilter.doFilterInternal(CsrfFilter.java:85)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:64)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:91)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:53)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:330)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:213)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:176)
	at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:346)
	at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:262)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:87)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.session.web.http.SessionRepositoryFilter.doFilterInternal(SessionRepositoryFilter.java:126)
	at org.springframework.session.web.http.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:65)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:121)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.springframework.boot.actuate.autoconfigure.MetricsFilter.doFilterInternal(MetricsFilter.java:103)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:212)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:141)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:521)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1096)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:674)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1500)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1456)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
````
The session is not being recognized after that and app's page says after that
````
There was an unexpected error (type=Forbidden, status=403).
Expected CSRF token not found. Has your session expired?
````

That looks like something unexpected for the spring-session code.
Am I missing anything, or this is a spring-session issue indeed?

**Comment by rwinch** [comment]
Thanks for the report!

I think @jxblum is the best person to look into this. What are your thoughts John?

**Comment by jxblum** [comment]
So, typically, "Tombstones" are not created by GemFire until after an entry is destroyed and are only required to enforce strict consistency (i.e. operation order) amongst multiple nodes in a cluster.

See the [Consistency Checking by Region Type](http://gemfire.docs.pivotal.io/docs-gemfire/latest/developing/distributed_regions/how_region_versioning_works.html), and specifically the section titled "_How Destroy and Clear Operations Are Resolved_",  in the GemFire User Guide for further details.

By default, I have [configured the PARTITION Region](https://github.com/spring-projects/spring-session/blob/master/spring-session/src/main/java/org/springframework/session/data/gemfire/config/annotation/web/http/GemFireHttpSessionConfiguration.java#L374-L375) into which (HTTP) Sessions are stored in GemFire to use an entry idle timeout of `maxInactiveInterval` with an action of `INVALIDATE`.  Therefore, if a user does nothing, the entry will eventually timeout and GemFire will mark the entry as invalid.  This is useful in cases where the application maybe accessing Sessions outside of the Spring Session API or implicitly during the normal operation of the application using Spring Session's infrastructure.

Logistically, Spring Session using GemFire will also [detect stale Sessions (i.e. Sessions that have "expired") and delete them upon retrieval](https://github.com/spring-projects/spring-session/blob/master/spring-session/src/main/java/org/springframework/session/data/gemfire/GemFireOperationsSessionRepository.java#L117-L121).  This in turn (indirectly) [triggers a `Region.remove(key)`](https://github.com/spring-projects/spring-data-gemfire/blob/1.7.x/src/main/java/org/springframework/data/gemfire/GemfireTemplate.java#L226) to occur [using the SDG GemfireTemplate](https://github.com/spring-projects/spring-session/blob/master/spring-session/src/main/java/org/springframework/session/data/gemfire/GemFireOperationsSessionRepository.java#L151).  Note, `Region.remove(key)` is just a [wrapper around `Region.destroy(key)`](https://github.com/apache/incubator-geode/blob/rel/v1.0.0-incubating.M1/gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/LocalRegion.java#L9768).

The interesting thing in this case is that the act of retrieving (i.e. `Region.get(key)` where key == (new) sessionId) the Session triggered a Region.put(key).... huh?!

```
...
    at com.gemstone.gemfire.internal.cache.LocalRegion.basicPutPart2(LocalRegion.java:5987)
    at com.gemstone.gemfire.internal.cache.ProxyRegionMap.basicPut(ProxyRegionMap.java:228)
    at com.gemstone.gemfire.internal.cache.LocalRegion.basicPutEntry(LocalRegion.java:5919)
    at com.gemstone.gemfire.internal.cache.LocalRegion.findObjectInSystem(LocalRegion.java:2808)
    at com.gemstone.gemfire.internal.cache.LocalRegion.nonTxnFindObject(LocalRegion.java:1464)
    at com.gemstone.gemfire.internal.cache.LocalRegionDataView.findObject(LocalRegionDataView.java:133)
    at com.gemstone.gemfire.internal.cache.LocalRegion.get(LocalRegion.java:1348)
    at com.gemstone.gemfire.internal.cache.LocalRegion.get(LocalRegion.java:1311)
    at com.gemstone.gemfire.internal.cache.LocalRegion.get(LocalRegion.java:1298)
    at com.gemstone.gemfire.internal.cache.AbstractRegion.get(AbstractRegion.java:321)
    at org.springframework.data.gemfire.GemfireTemplate.get(GemfireTemplate.java:161)
    at org.springframework.session.data.gemfire.GemFireOperationsSessionRepository.getSession(GemFireOperationsSessionRepository.java:117)
```

It seems the Session would(/should) not exist yet since you mention are you are performing a login attempt. (?)  Also, it is the first time we passed through the Spring Session "filter" on this request...

```
at org.springframework.session.web.http.OncePerRequestFilter.doFilter([OncePerRequestFilter.java:65](https://github.com/spring-projects/spring-session/blob/master/spring-session/src/main/java/org/springframework/session/web/http/OncePerRequestFilter.java#L65))
```
Anyway, to work around this problem you can disable GemFire's concurrency checking by setting the `concurrency-checks-enabled` to *false* on the (Session) Region (attributes), like so...

```
    @Bean
    BeanPostProcessor disableConcurrencyChecks() {
      return new BeanPostProcessor() {
        public Object postProcessBeforeInitialization(final Object bean, final String beanName) throws BeansException {
          if (bean instanceof RegionAttributesFactoryBean && "sessionRegionAttributes".equals(beanName)) {
            ((RegionAttributesFactoryBean) bean).setConcurrencyChecksEnabled(false);
          }

          return bean;
        }

        public Object postProcessAfterInitialization(final Object bean, final String beanName) throws BeansException {
          return bean;
        }
      };
    }
```

As indicated on the same page noted above in the _GemFire User Guide_, but under the section "_Configuring Consistency Checking_", consistency checks are enabled by default.  Strict consistency might be necessary if the user is attempting to reuse the same (authenticated) Session across multiple applications.

**Comment by jxblum** [comment]
@sshcherbakov @rwinch Also note, I just created a separate branch ([gh-373](https://github.com/jxblum/spring-session/tree/gh-373)) in my [fork of _Spring Session_](https://github.com/jxblum/spring-session) that [modifies the 'Users' sample application](https://github.com/jxblum/spring-session/commit/caa154b68fc0b7c08e12b7870adddd3eba71b4f5) as well as the ['Security' sample application](https://github.com/jxblum/spring-session/commit/23b17698d1ca2de1d820410dbeef27091edeb5cc) to use _GemFire_ as the backing store.

Both samples worked successfully.

@sshcherbakov Would it be possible for you to create a simple, single, self-contained/standalone application (or test) that demonstrates this problem?

**Comment by jxblum** [comment]
@sshcherbakov Concerning why Spring Session Data GemFire support uses "DataSerialization" vs. PDX... well, that boiled to the fact that PDX has limited capabilities for handling deltas, and the way in which I wanted to "serialize" Session attributes (effectively treating them as first-class 'properties' of the Session object itself for convenience in indexing & querying) does not work when the Session attributes are dynamic (can be added @ runtime) since the PDX type meta-data for the Session class is fixed.  I.e. you cannot store additional Session attributes in the Session once the Session has already been serialized using PDX.  You essentially have to reconstitute (reconstruct) the PDX meta-data for the Session object in it's current state.

**Comment by sshcherbakov** [comment]
Thank you @jxblum 
I've been using replicated region indeed.
It should be similar to partitioned with redundancy added.
I'll try the pure partitioned now (and switching off consistency checks and trying to simplify the sample project for investigation).

**Comment by sshcherbakov** [comment]
I've checked quickly.
I see same error when the region is pure partitioned (--type=PARTITION)
Also the BeanProcessor that disables the concurrency checks doesn't help (though get's called for sure.
````
gfsh>describe region --name=/ClusteredSpringSessions
..........................................................
Name            : ClusteredSpringSessions
Data Policy     : partition
Hosting Members : server1

Non-Default Attributes Shared By Hosting Members

 Type  | Name | Value
------ | ---- | -----
Region | size | 1
````
````
2016-02-23 09:49:05.252 DEBUG 5947 --- [nio-8080-exec-3] o.s.b.a.audit.listener.AuditListener     : AuditEvent [timestamp=Tue Feb 23 09:49:05 CET 2016, principal=user, type=AUTHENTICATION_SUCCESS, data={details=org.springframework.security.web.authentication.WebAuthenticationDetails@21a2c: RemoteIpAddress: 0:0:0:0:0:0:0:1; SessionId: 9c84b5a7-edb8-4826-86d1-484d948054df}]
[error 2016/02/23 09:49:05.258 CET <http-nio-8080-exec-3> tid=0x45] Exception occurred in CacheListener
java.lang.ClassCastException: com.gemstone.gemfire.internal.cache.Token$Tombstone cannot be cast to org.springframework.session.ExpiringSession
	at org.springframework.session.data.gemfire.AbstractGemFireOperationsSessionRepository.afterCreate(AbstractGemFireOperationsSessionRepository.java:212)
	at com.gemstone.gemfire.internal.cache.EnumListenerEvent$AFTER_CREATE.dispatchEvent(EnumListenerEvent.java:97)
````

I'm wondering whether the way I create the region matters.
I'm doing that manually dynamically from the gfsh command line before starting my app (otherwise the app doesn't find the region).
I'll share with you the link to my test project in the current state.
(And then will continue tests a bit later)

**Comment by jxblum** [comment]
@rwinch @sshcherbakov I just retested the [Spring Session Security sample application](http://docs.spring.io/spring-session/docs/1.1.0.RC1/reference/html5/guides/security.html) [with *GemFire*](https://github.com/jxblum/spring-session/blob/gh-373/samples/security/src/main/java/sample/GemFireConfig.java) using a `REPLICATE` Region, setting the `maxInactiveIntervalInSeconds` to *30*, allowing the user's Session to timeout more quickly and thus force me to have to re-login sooner.  Hence...

```
@EnableGemFireHttpSession(serverRegionShortcut = RegionShortcut.REPLICATE, maxInactiveIntervalInSeconds = 30)
public class GemFireConfig {
```

The idea behind this configuration/setup is that GemFire should leave behind a "_Tombstone_" for the previous Session, so when I try to log back in, if there is any anomaly with (Session) data stored in the Region that hopefully I would encounter it upon re-login.

However, once again, the Security sample application worked as expected.

I am now going spend some cycles looking into the https://github.com/pdeemea/daimler-udc app.

Stay tuned...

**Comment by jxblum** [comment]
@sshcherbakov Just sending some quick observations...

1. Technically, the `disableConcurrencyChecks *BeanPostProcessor*` should be declared/registered on the server.

Of course, that assumes you are configuring and bootstrapping your _GemFire Server_ with _Spring_, such as by using _Spring Boot_ along with _Spring (Data GemFire)_ Java or XML-based configuration meta-data and _Spring Session's_ `@EnableGemFireHttpSession` annotation (or alternatively, [creating a bean of type `GemFireHttpSessionConfiguration` in XML](https://github.com/spring-projects/spring-session/blob/master/spring-session/src/test/resources/org/springframework/session/data/gemfire/config/annotation/web/http/GemFireHttpSessionXmlConfigurationTests-context.xml#L21-L23)); for example...

```
@SpringBootApplication
@EnableGemFireHttpSession(maxInactiveIntervalInSeconds = 30)
public class SpringSessionGemFireServerApplication {

    public static void main(final String[] args) {
        ...
    }
    
    // GemFire configuration omitted...
}
````

In this case, the `@EnableGemFireHttpSession` annotation conveniently sets up the server with the appropriate configuration for you... the `ClusteredSpringSessions` PARTITION _Region_, expiration policy, indexes and so on, which must exist even when the topology used by your application to maintain clustered, replicated Sessions is client/server based.  This is even more convenient than using _Gfsh_, to setup objects in _GemFire_, required by _Spring Session_, manually.

However, it does require the use of a _Spring_ `BeanPostProcessor` to get at and modify the _GemFire_ configuration automatically setup by _Spring Session's_ `@EnableGemFireHttpSession` annotation.

Of course, you can also "override" the _Spring Session_, _GemFire_ bean definitions providing you know the names of the beans.  So, you have many choices, but I will not discuss bean definition overriding here.

2.  If you are configuring and starting a _GemFire Server_ manually, with _Spring Boot_, _Gfsh_ (as you have done) or otherwise, then again, you have many options for disabling **concurrency checking**.

The following approaches are applicable when you are **NOT** conveniently and implicitly configuring the _GemFire Server_ with the _Spring Session_ `@EnableGemFireHttpSession` annotation as discussed above.

If you are using _Spring (Data GemFire)_ Java or XML-based config, then you can just do the following...

```
<!-- Spring Session Data GemFire Region definition on the server configured using Spring XML -->
<gfe:partitioned-region name="ClusteredSpringSessions" persistent="false" concurrency-checks-enabled="false"/>
```

Or by way of _Gfsh_...

```
gfsh>create region --name=ClusteredSpringSessions --type=PARTITION --enable-concurrency-checks=false
```

3. Note, after starting your front-end web application, **daimler-udc**, I noticed the following on the client...

```
[info 2016/02/23 19:51:21.322 PST <Cache Client Updater Thread  on jblum-mbpro-2(SpringSessionGemFireServerApplication:29982)<v0>:51997 port 40404> tid=0x1e] Server has concurrencyChecksEnabled true but client has false for region com.gemstone.gemfire.internal.cache.LocalRegion[path='/ClusteredSpringSessions';scope=LOCAL';dataPolicy=EMPTY]
```

This is clearly indicates the _Spring_ `BeanPostProcessor` has taken affect on the client, but is not (properly) configured on the server, where it matters.  I.e. the server should have `concurrencyChecksEnabled` set to **false** as well.

4.  Interestingly enough, I successfully started your application, left it sit idle as I was typing this, went back to attempt a first-time login, then hit the following error upon login...

```
Whitelabel Error Page

This application has no explicit mapping for /error, so you are seeing this as a fallback.

Tue Feb 23 20:26:05 PST 2016
There was an unexpected error (type=Forbidden, status=403).
Expected CSRF token not found. Has your session expired?
```

This error does correspond to the Session timeout, but is **NOT** correlated with the `ClassCastException` from _GemFire_....

```
[error 2016/02/22 02:29:59.823 CET <http-nio-8080-exec-2> tid=0x31] Exception occurred in CacheListener
java.lang.ClassCastException: com.gemstone.gemfire.internal.cache.Token$Tombstone cannot be cast to org.springframework.session.ExpiringSession
```

I.e. the _GemFire_ `ClassCastException` has not occurred yet, even when the unauthenticated (GemFire) Session times out.  My guess is, your application does not properly handle the Session timeout when presenting the error page, possibly with information from the now "expired" Session.

5.  After hitting the "Whitelabel Error Page", if I just restart from http://localhost:8080 in my browser, then the login page comes back up.  I am guessing your /error resource mapping is incorrect, however, it has nothing to do with the GemFire error, which is technically benign and misleading.

6. So... I configured your application to connect to **Redis** instead of _GemFire_ and I ran into the same problem!

7. I also hit this "_Whitelabel Error Page_" by **NOT** selecting a file first and clicking the "Upload" button, after successfully logging in.

8. In fact, I just hit the "_Whitelabel Error Page_" even when I "Choose File", specify a "Name" and click the "Upload" button!  NOTE: this is using the `LocalFileService` configuration.

9. Anyway... if I go back to running your application with GemFire managing the Spring Sessions, I login and wait for the Session to expire, when the `concurrencyChecksEnabled` is appropriately configured to **false** on the server, I **NO** longer see the `ClassCastException`, as I would expect.  If I login, attempt to upload a file, I get the same "_Whitelabel Error Page_" as with Redis, however, I can reset the upload file form page by entering http://localhost:8080/, which indicates that my "authenticated" (GemFire) Session is still intact.

In summary... outside of the benign and misleading _GemFire_ `ClassCastException`, I would say _Spring Session_ using _GemFire_ support is functioning exactly as it should, or at least, consistently with Redis support enabled instead.

I also have a theory for why _GemFire_ is also doing a `Region.put(sessionId)` operation (technically on the client) with the "expired" Session (which is now a "_Tombstone_" on the server after the Session entry times out/expires, which is what the server ends up propagating to the client)... it is so the client (PROXY) Region CacheListener is invoked to receive the event.

Finally, I made some changes to your **daimler-udc** project.  I created the _gh-373_ branch and [submitted a PR](https://github.com/pdeemea/daimler-udc/pull/1) for your reference.

**Comment by jxblum** [comment]
@rwinch I decided to do some (_Spring Session Data GemFire_ support) cleanup based on @sshcherbakov findings.

In this case, I am appropriately suppressing any _GemFire_ (cache Region) event related Exceptions (such as the `ClassCastException` caused by an "expired" Session stored in _GemFire_ on the server, leaving behind a "_Tombstone_" when _concurrency checks_ are **enabled** for consistency sake).  In addition as well as ensuring that Session events (e.g. SessionCreatedEvent) are only fired when it is necessary (i.e. when GemFire passes the "_Tombstone_" entry to the client, which results in a `Region.put(..)` operation in order to trigger the event on the client, it is NOT necessary to fire a SessionCreatedEvent in this case).

**Comment by jxblum** [comment]
At this point, I consider this issue closed.

**Comment by sshcherbakov** [comment]
Thank you @jxblum !

Setting the `--enable-concurrency-checks=false` on the server side fixed the ClassCastException problem. I hope that disabling consistency check won't remain requirement for using spring-session eventually though.

The error during authentication was caused by my incorrect spring-security configuration.
It all works very well now!

**Comment by jxblum** [comment]
That is great news! Awesome @sshcherbakov!

We can revisit concurrency checks later if necessary and I will put more thought into it on the Spring Session (Data GemFire) side as well.

Cheers!
